### PR TITLE
fix(asgi): handle invalid headers

### LINKED ISSUE #2103: UnicodeDecodeError for invalid headers in ASGI middleware
When the ASGI middleware receives headers with invalid encoding, for example as the result of somebody fuzzing an endpoint as is the case here, it introduces a Unicode decode error into the HTTP pipeline. It shouldn't be doing this.

### Which version of dd-trace-py are you using?

0.46.0

### Which version of the libraries are you using?

What I believe are the potentially relevant libraries are listed below.

```
datadog==0.40.1
...
ddtrace==0.46.0
...
fastapi==0.63.0
...
httpcore==0.12.3
httptools==0.1.1
...
starlette==0.13.6
...
uvicorn==0.13.4
uvloop==0.15.2
```

### How can we reproduce your problem?

```python
from ddtrace.contrib.asgi.middleware import _extract_headers


def test_ddtrace_invalid_header_name_encoding():
    assert _extract_headers({"headers": [((bytes.fromhex("c0")), "test")]}) == {}


def test_ddtrace_invalid_header_value_encoding():
    assert _extract_headers({"headers": [("test", (bytes.fromhex("c0")))]}) == {}
```

### What is the result that you get?

The middleware blows up with a `UnicodeDecodeError: 'utf-8' codec can't decode byte 0xc0 in position 0: invalid start byte` exception.

### What is the result that you expected?

This is trace middleware so it should not introduce exceptions into request processing. In the case of a header it can't decode then I'd expect it to either:

1. Omit the invalid header from the result and continue attempting to decode the remainder, or
2. Substitute a sentinel value for the unprocessable header to indicate an error

I've suggested (1) in the repro above, as that seems like it'd be the least likely to cause any problems.

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
# Bug fix

## Description

Reported in #2103.

The ASGI specification on headers:

> headers: These are byte strings of the exact byte sequences sent by the client/to be sent by the server. While modern HTTP standards say that headers should be ASCII, older ones did not and allowed a wider range of characters. Frameworks/applications should decode headers as they deem appropriate.

## Fix

This patch adds error handling for when headers cannot be decoded.

Fixes #2103

# Checklist
- [x] Regression test added; or
- [x] Description of manual testing performed and explanation is included in the code and/or PR.
- [x] Added to the correct milestone.

### CODE REVIEW

File: ddtrace/contrib/asgi/middleware.py [hunk 1]
```diff
@@ -85,11 +85,15 @@ async def __call__(self, scope, receive, send):
...
 
-        headers = _extract_headers(scope)
-
-        trace_utils.activate_distributed_headers(
-            self.tracer, int_config=self.integration_config, request_headers=headers
-        )
+        try:
+            headers = _extract_headers(scope)
+        except Exception:
+            log.warning("failed to decode headers", exc_info=True)
```
Comment (brettlangdon) [inline ddtrace/contrib/asgi/middleware.py:91]: should we make this more actionable?

e.g. `failed to decode headers, cannot parse distributed tracing headers for request` ?

Comment (brettlangdon) [reply ddtrace/contrib/asgi/middleware.py:91]: "actionable"

Comment (Kyle-Verhoog) [reply ddtrace/contrib/asgi/middleware.py:91]: :+1: yes! This log message sucked lol

**Review by majorgreys — APPROVED** [review]

**Review by brettlangdon — CHANGES_REQUESTED** [review]
once log message is updated, then lgtm

**Comment by Kyle-Verhoog** [comment]
It also looks like the test case doesn't work properly because `httpx` itself doesn't support decoding the headers used in our test case....

**Review by brettlangdon — APPROVED** [review]

**Review by mabdinur — APPROVED** [review]
LGTM

**Review by mabdinur — CHANGES_REQUESTED** [review]
I missed this test in my review. tests.contrib.django.test_django_snapshots.test_404_exceptions_18x

**Comment by Kyle-Verhoog** [comment]
@mabdinur it's a flaky test - unrelated to this change

**Review by mabdinur — APPROVED** [review]
LGTM

**Comment by majorgreys** [comment]
@Mergifyio backport 0.51

**Pull request merged**